File: gamelogic/src/main/java/com/alta_v2/game/gamelogic/stage/MapStage.java
package com.alta_v2.game.gamelogic.stage;

import com.alta_v2.facade.dialogApi.DialogCoreApi;
import com.alta_v2.game.gamelogic.data.map.MapModel;
import com.alta_v2.game.gamelogic.data.npc.NpcModel;
import com.alta_v2.game.gamelogic.domain.npc.RepeatableActionProcessor;
import com.alta_v2.game.gamelogic.domain.player.PlayerProcessor;
import com.alta_v2.game.gamelogic.stage.event.ChangeMapStageEvent;
import com.google.inject.assistedinject.Assisted;
import com.google.inject.assistedinject.AssistedInject;
import lombok.extern.log4j.Log4j2;

import java.util.List;

@Log4j2
public class MapStage extends AbstractStage {

    private final PlayerProcessor playerProcessor;
    private final RepeatableActionProcessor npcProcessor;
    private final DialogCoreApi dialogCoreApi;
    private final MapModel mapModel;

    @AssistedInject
    public MapStage(@Assisted MapModel mapModel,
                    @Assisted List<NpcModel> npcList,
                    PlayerProcessor playerProcessor,
                    RepeatableActionProcessor npcProcessor,
                    DialogCoreApi dialogCoreApi) {
        this.playerProcessor = playerProcessor;
        this.npcProcessor = npcProcessor;
        this.dialogCoreApi = dialogCoreApi;
        this.mapModel = mapModel;

        npcList.forEach(this.npcProcessor::addToRandomMovement);
        this.npcProcessor.startAsync();
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void onActionBegin(ActionType action) {
        this.playerProcessor.actionBegin(action);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void onActionFinish(ActionType action) {
        this.playerProcessor.actionFinish(action);

        if (action == ActionType.BACK) {
            dialogCoreApi.hideTitleDialog();
            this.changeStage(new ChangeMapStageEvent());
        }
    }

    @Override
    public void onStageLoaded() {
        dialogCoreApi.showTitleDialog(mapModel.getDisplayName(), 3000);
    }

    /**
     * {@inheritDoc}
     */
    @Override
    public void destroy() {
        super.destroy();
        this.playerProcessor.destroy();
        this.npcProcessor.destroy();
    }
}

File: gamelogic/src/main/kotlin/com/alta_v2/game/gamelogic/stage/MapStage.kt
package com.alta_v2.game.gamelogic.stage

import com.alta_v2.game.gamelogic.data.map.MapModel
import com.alta_v2.game.gamelogic.data.npc.NpcModel
import com.alta_v2.game.gamelogic.interaction.InteractionFactory
import com.alta_v2.game.gamelogic.interaction.InteractionManager
import com.alta_v2.game.gamelogic.utils.LogicThreadFactory
import com.alta_v2.mediator.serde.ActionType
import com.google.common.collect.Sets
import com.google.inject.assistedinject.Assisted
import com.google.inject.assistedinject.AssistedInject
import kotlinx.coroutines.*
import mu.KotlinLogging
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit

private const val LISTENER_THREAD_NAME = "map-stage-listener"
private const val ACTION_THREAD_NAME = "map-stage-action-worker"

class MapStage @AssistedInject constructor(@Assisted npcList: List<NpcModel>,
                                           @Assisted mapModel: MapModel,
                                           interactionFactory: InteractionFactory) : AbstractStage() {

    private val log = KotlinLogging.logger {  }
    private val actionStatus = Sets.newConcurrentHashSet<ActionType>()
    private val interactionManager: InteractionManager = interactionFactory.createInteractionManager(mapModel, npcList)
    private val repeatableActionConsumer = {
        actionType: ActionType -> interactionManager.onRepeatableAction(actionType)
    }

    @ObsoleteCoroutinesApi
    private val actionExecutor = newFixedThreadPoolContext(5, ACTION_THREAD_NAME)
    private val scheduledExecutor = Executors.newScheduledThreadPool(1, LogicThreadFactory(LISTENER_THREAD_NAME))


    init {
        scheduledExecutor.scheduleWithFixedDelay(::handleRepeatableActions, 0L, 10L, TimeUnit.MILLISECONDS)
        interactionManager.getChangeMapFuture().thenAccept(::changeStage)
    }

    override fun onActionBegin(action: ActionType) {
        actionStatus.add(action)
    }

    @ObsoleteCoroutinesApi
    override fun onActionFinish(action: ActionType) {
        actionStatus.remove(action)

        CoroutineScope(actionExecutor).launch {
            interactionManager.onPress(action)
        }
    }

    override fun onStageLoaded() {
        interactionManager.onInitialized()
    }

    override fun destroy() {
        super.destroy()
        interactionManager.destroy()

        try {
            scheduledExecutor.shutdown()
        } catch (e: Exception) {
            log.error("Failed to shutdown scheduled executor service", e)
        }
    }

    private fun handleRepeatableActions() {
        try {
            actionStatus.isNotEmpty().run { actionStatus.forEach(repeatableActionConsumer) }
        } catch (e: Exception) {
            log.error("Failed to handle repeatable actions", e)
        }
    }
}